test/scale/run_test.py

```diff
@@ -0,0 +1,372 @@
+            else:
+                tmp_buf.append(char)
+        self.identity = "".join(tmp_buf)
+    
+    def reset(self):
+        self.latencies = []
+
+    def add_latency(self, latency):
+        self.latencies.append(latency)
+    
+    def get_result(self):
+        if len(self.latencies) == 0:
+            self.latencies.append(0)
+        return {
+            "func_name": self.FUNC_NAME,
+            "num": len(self.latencies),
+            "min": min(self.latencies),            
+            "max": max(self.latencies),
+            "avg": round(sum(self.latencies) / len(self.latencies), 4),
+            "points": self.latencies
+        }
+
+
+class FuncControllerCreateVolume(FuncPerfResult):
+    FUNC_NAME = "ControllerCreateVolume"
+
+
+class FuncControllerDeleteVolume(FuncPerfResult):
+    FUNC_NAME = "ControllerDeleteVolume"
+
+
+class FuncControllerPublishVolume(FuncPerfResult):
+    FUNC_NAME = "ControllerPublishVolume"
+
+
+class FuncControllerUnpublishVolume(FuncPerfResult):
+    FUNC_NAME = "ControllerUnpublishVolume"
+
+
+class FuncNodePublishVolume(FuncPerfResult):
+    FUNC_NAME = "NodePublishVolume"
+
+
+class FuncNodeUnpublishVolume(FuncPerfResult):
+    FUNC_NAME = "NodeUnpublishVolume"
+
+
+class PerfResultCollector:
+    def __init__(self):
+        self._perf_identity = re.compile(r"latency_seconds=([0-9.e-]*)")
+        self._func_data = {}
+
+    def register_func(self, func):
+        self._func_data.setdefault(func.identity, func)
+
+    def parse_log_file(self, f):
+        for line in f:
+            match = self._perf_identity.search(line)
+            if match is not None:
+                for func_name, func in self._func_data.items():
+                    if line.find(func_name) != -1:
+                        func.add_latency(float(match.group(1)))
+                        break
+
+    def get_result(self):
+        return {
+            "func_results": [
+                func.get_result() for func in self._func_data.values()
+            ]
+        }
+
+
+class PerfScaleTest:
+    WORK_LOAD_YAML_PATH = ""
+    TARGET_FUNCS = []
+
+    def __init__(self, args):
+        self.configurations = {
+            "csi_name": args.csi_name,
+            "mgs_ip_address": args.mgs_ip_address,
+            "fs_name": args.fs_name,
+            "scale": 0,
+        }
+        self._scales = args.scales
+        logger.info(f"test scales {self._scales}")
+        self._test_result = None
+        self._provisioning_type = args.provisioning_type
+        self._generated_workload_yaml = os.path.join(WORK_PATH,
+                                                     "tmp_workload.yml")
+        self._csi_log_path = os.path.join(WORK_PATH, "logs")
+        self._result_path = WORK_PATH
+        logger.info(f"template workload yaml {self.WORK_LOAD_YAML_PATH} "
+                    f"result path {self._result_path}")
+
+        self._current_scale = 0
+
+    def generate_workload_yaml(self):
```

Comment:
Is there a python library that can generate yaml ?

Reply:
Hi @joe-atzinger. There are many libraries can parse template files, like jinjia2. But I'm trying to avoid using any third part libraries. So that, I can skip setting up the virtual env and installing python requirements in pipeline. Fortunately, this template and replacement is simple.